expression/builtin_cast.go

```diff
@@ -829,12 +828,16 @@ type builtinCastDecimalAsDurationSig struct {
 }
 
 func (b *builtinCastDecimalAsDurationSig) evalDuration(row []types.Datum) (res types.Duration, isNull bool, err error) {
-	val, isNull, err := b.args[0].EvalDecimal(row, b.getCtx().GetSessionVars().StmtCtx)
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	val, isNull, err := b.args[0].EvalDecimal(row, sc)
 	if isNull || err != nil {
-		return res, false, errors.Trace(err)
+		return res, true, errors.Trace(err)
 	}
-	res, err = types.ParseDuration(string(val.ToString()), b.tp.Decimal)
-	return res, false, errors.Trace(err)
+	res, isNull, err = types.ParseDuration(string(val.ToString()), b.tp.Decimal)
+	if types.ErrTruncatedWrongVal.Equal(err) {
+		err = sc.HandleTruncate(err)
+	}
+	return
```

Comment:
```go
return res, isNull || err != nil, errors.Trace(err)
```

Reply:
sc.HandleTruncate check is need to deal with SQLMODE.